squeal-postgresql/src/Squeal/PostgreSQL/Query/With.hs

```diff
@@ -161,5 +185,56 @@ instance
 
 instance (forall c s p r. RenderSQL (statement c s p r)) => RenderSQL
   (CommonTableExpression statement db params with0 with1) where
-    renderSQL (CommonTableExpression (statement `As` cte)) =
-      renderSQL cte <+> "AS" <+> parenthesized (renderSQL statement)
+    renderSQL (CommonTableExpression (statement `As` cte) materialization) =
+      renderSQL cte
+        <+> "AS"
+        <+> renderSQL materialization
+        <> case materialization of
+              DefaultMaterialization -> ""
+              _ -> " "
+        <> parenthesized (renderSQL statement)
+
+{- | Force separate calculation of the WITH query.
+
+>>> :{
+let
+  qry :: Query lat with (Public Schema) params '["col1" ::: 'NotNull 'PGint4, "col2" ::: 'NotNull 'PGint4]
+  qry = with (
+    materialized (select Star (from (table #tab)) `as` #cte1) :>>
+    select Star (from (common #cte1)) `as` #cte2
+    ) (select Star (from (common #cte2)))
+in printSQL qry
+:}
+WITH "cte1" AS MATERIALIZED (SELECT * FROM "tab" AS "tab"), "cte2" AS (SELECT * FROM "cte1" AS "cte1") SELECT * FROM "cte2" AS "cte2"
+
+Note: if the last CTE has `materialized` or `notMaterialized` you must add `:>> Done`.
+
+Requires PostgreSQL 12 or higher.
+-}
+materialized
+  :: CommonTableExpression statement db params with with1
```

Comment:
should the input type be `Aliased (statement db params with) common` instead? Otherwise it's not necessarily clear what `materialized . notMaterialized` would mean.

Reply:
Yeah sounds good to me. 4c9b44f